### PR TITLE
feat: add swap component

### PR DESCRIPTION
**Pull request by MishaZhem**
Closes #

### CODE REVIEW

File: apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.html [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+                        [formControl]="swapForm['controls'][$index]"
+                        (change)="newSwap(data, $index)"
+                    />
+                </tui-input-inline>
+                <div>
+                    <div
+                        tuiSubtitle
+                        class="equal-number"
+                    >
+                        ≈{{ getPrice(data, swapService[$index].chosen, swapForm.controls[$index].value) }}$
```
Comment (vladimirpotekhin) [inline apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.html:78]: let's use tuiAmount pipe

File: apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.html [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+                    </div>
+                    <hr />
+                </div>
+            </div>
+        }
+        <button
+            appearance="primary"
+            tuiButton
+            (click)="newSwap(data, 0)"
```
Comment (vladimirpotekhin) [inline apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.html:87]: why second arg 0?

File: apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.html [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+                    tuiButton
+                    tuiChevron
+                    type="button"
+                    class="choose-crypto"
+                    [tuiDropdown]="dropdownContent"
+                    [tuiDropdownManual]="swapService[$index].status"
+                    (click)="openInfo($index)"
+                >
+                    <tui-avatar
+                        size="s"
```
Comment (vladimirpotekhin) [inline apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.html:33]: too big avatar, try to use xs size

File: apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.html [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+                    <div TuiTitle>{{ type.title }}</div>
+                    <div TuiTitle>Balance: 1000 {{ swapService[$index].chosen.toUpperCase() }}</div>
+                </div>
+                <button
+                    tuiButton
+                    tuiChevron
+                    type="button"
+                    class="choose-crypto"
+                    [tuiDropdown]="dropdownContent"
+                    [tuiDropdownManual]="swapService[$index].status"
```
Comment (vladimirpotekhin) [inline apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.html:29]: dropdown should close after selection and after click outside the list. Try to use DropdownOpen (first example) https://taiga-ui.dev/next/directives/dropdown-open

File: apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.service.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
+import {Injectable} from '@angular/core';
+
+interface SwapData {
+    readonly title: string;
+    status: boolean;
+    chosen: string;
```
Comment (vladimirpotekhin) [inline apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.service.ts:6]: what is status 'false' ? 

avoid to mutate initial data, all data fields should be readonly

Comment (vladimirpotekhin) [reply apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.service.ts:6]: and remove "choosen" from interface, it's the current selection state not part of the data

File: apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.ts [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    ],
+    templateUrl: './swap.component.html',
+    styleUrl: './swap.component.less',
+    changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class SwapComponent {
+    protected cryptoService = inject(CryptoService);
+    protected info$ = this.cryptoService.info$;
+    protected swapService = inject(SwapService).swapData;
+    protected swapForm = new FormArray([new FormControl(0.22), new FormControl(0.22)]);
```
Comment (vladimirpotekhin) [inline apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.ts:63]: let's use 0 for initial values

File: apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.ts [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        for (const token of data) {
+            if (token && token.symbol.toLowerCase() === title.toLowerCase()) {
+                return (Number(token.priceUsd) * value).toFixed(2);
+            }
+        }
+
+        return value.toFixed(2);
+    }
+
+    protected newSwap(data: PricesData[], current: number): void {
```
Comment (vladimirpotekhin) [inline apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.ts:88]: let's try more declatative approach

File: apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.ts [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+import type {PricesData} from '../../../../services/crypto.service';
+import {CryptoService} from '../../../../services/crypto.service';
+
+@Component({
+    standalone: true,
+    selector: 'lmb-swap',
+    imports: [
+        CommonModule,
+        FormsModule,
+        ReactiveFormsModule,
```
Comment (vladimirpotekhin) [inline apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.ts:40]: ReactiveFormsModule unused, check others imports as well

File: apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.ts [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+        TuiTextfieldControllerModule,
+        TuiTitle,
+    ],
+    templateUrl: './swap.component.html',
+    styleUrl: './swap.component.less',
+    changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class SwapComponent {
+    protected cryptoService = inject(CryptoService);
+    protected info = toSignal(this.cryptoService.getTokens());
```
Comment (vladimirpotekhin) [inline apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.ts:69]: protected readonly tokens = toSignal(this.cryptoService.getTokens());

Comment (vladimirpotekhin) [reply apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.ts:69]: check if you can make other fields readonly

File: apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.ts [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+    protected priceFrom = computed(() => this.getPrice(this.info(), this.chosen()[0]));
+
+    protected priceTo = computed(() => this.getPrice(this.info(), this.chosen()[1]));
+
+    protected titles = ['From', 'To'];
+
+    protected from = signal('0');
+    protected to = signal('0');
+    protected chosen = signal(['eth', 'btc']);
```
Comment (vladimirpotekhin) [inline apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.ts:78]: lets make two separate fields - tokenFrom and tokenTo

File: apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.ts [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+    protected openedDialog = [false, false];
+    protected val = 0;
+
+    protected newToken(index: number, title: string): void {
+        this.chosen()[index] = title;
+        this.openedDialog[index] = false;
+    }
+
+    protected toNum(val: string): number {
```
Comment (vladimirpotekhin) [inline apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.ts:87]: remove this

File: apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.ts [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+    protected priceFrom = computed(() => this.getPrice(this.info(), this.chosen()[0]));
+
+    protected priceTo = computed(() => this.getPrice(this.info(), this.chosen()[1]));
+
+    protected titles = ['From', 'To'];
+
+    protected from = signal('0');
+    protected to = signal('0');
+    protected chosen = signal(['eth', 'btc']);
+    protected openedDialog = [false, false];
```
Comment (vladimirpotekhin) [inline apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.ts:79]: remove this

File: apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.ts [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+    protected val = 0;
```
Comment (vladimirpotekhin) [inline apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.ts:80]: what is this for?

File: apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.ts [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+        this.chosen()[index] = title;
+        this.openedDialog[index] = false;
+    }
+
+    protected toNum(val: string): number {
+        return Number(val);
+    }
+
+    protected getPrice(data: PricesData[] | undefined, title: string): number {
+        for (const token of data ?? []) {
```
Comment (vladimirpotekhin) [inline apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.ts:92]: `return +data.find(token => token.symbol.toLowerCase() === title.toLowerCase)?.priseUsd || 0`

File: apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.html [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+                    <div TuiTitle>{{ title }}</div>
+                    <div TuiTitle>Balance: 1000 {{ chosen()[$index].toUpperCase() }}</div>
+                </div>
+                <button
+                    tuiButton
+                    tuiChevron
+                    type="button"
+                    class="choose-crypto"
+                    [tuiDropdown]="dropdownContent"
+                    [(tuiDropdownOpen)]="openedDialog[$index]"
```
Comment (vladimirpotekhin) [inline apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.html:28]: just  `tuiDropdownOpen`
no need to manually set open/close state

File: apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.html [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+                    (click)="openedDialog[$index] = !openedDialog[$index]"
```
Comment (vladimirpotekhin) [inline apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.html:29]: remove this

File: apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.ts (hunk 12 above)
Comment (MishaZhem) [reply apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.ts]: openedDialog?

File: apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.ts [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+
+    protected readonly priceTo = computed(() =>
+        this.getPrice(this.info(), this.chosenTo()),
+    );
+
+    protected readonly titles = ['From', 'To'];
+
+    protected readonly from = signal(0);
+    protected readonly to = signal(0);
+    protected readonly chosenFrom = signal('eth');
```
Comment (vladimirpotekhin) [inline apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.ts:65]: chosen what? code should be self-documenting, let's use names that reflect entities
priceFrom, tokenFrom etc

Снимок экрана 2024-08-08 в 09 59 57

File: apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.ts [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+    ],
+    templateUrl: './swap.component.html',
+    styleUrl: './swap.component.less',
+    changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class SwapComponent {
+    protected readonly cryptoService = inject(CryptoService);
+    protected readonly info = toSignal(this.cryptoService.getTokens());
+    protected readonly priceFrom = computed(() =>
+        this.getPrice(this.info(), this.chosenFrom()),
```
Comment (vladimirpotekhin) [inline apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.ts:54]: this.getPrice(this.tokenFrom())
....

getPrice(token:string) {
  return (this.tokens() ?? []).find(....)
}

File: apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.html [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+                            [(ngModel)]="to"
+                            (change)="newSwapTo()"
+                        />
+                    } @else {
+                        <input
+                            max="1000"
+                            min="0"
+                            type="number"
+                            [(ngModel)]="from"
+                            (change)="newSwapFrom()"
```
Comment (vladimirpotekhin) [inline apps/taiga-lumbermill/src/dashboards/crypto/components/swap/swap.component.html:77]: change event

**Review by vladimirpotekhin — APPROVED** [review]

**Pull request merged**